metadetect/lsst/util.py

```diff
@@ -536,3 +536,217 @@ def trim_odd_image(im):
+    and the weight is converted to variance.
+
+    If a complex WCS or PSF is needed, it is better to send an existing one.
+    The wcs will not be altered by this code.
+
+    The psf image can be optionally copied in as a KernelPsf
+
+    Parameters
+    ----------
+    obs: ngmix.Observation
+        The observation data
+    exp: lsst.afw.image.ExposureF, optional
+        Optional exposure into which the data are to copied.  Note the psf is
+        not copied unless copy_psf is set to True.  The wcs is not modified.
+    copy_psf: bool, optional
+        If set to True, the psf is copied in as a KernelPsf(FixedKernel)
+
+    Returns
+    -------
+    exp: lsst.afw.image.ExposureF
+        Either a reference to the input exposure, or a new one if it was
+        created internally.
+    """
+    from lsst.meas.algorithms import KernelPsf
+    from lsst.afw.math import FixedKernel
+    import lsst.afw.image as afw_image
+
+    if exp is None:
+        ny, nx = obs.image.shape
+        exp = afw_image.ExposureF(nx, ny)
+
+    exp.image.array[:, :] = obs.image
+
+    exp.variance.array[:, :] = np.inf
+
+    if copy_mask_from == 'bmask':
+        mask = obs.bmask
+    elif copy_mask_from == 'ormask':
+        mask = obs.ormask
+    else:
+        raise RuntimeError(
+            f"copy_mask_from '{copy_mask_from}' should be 'bmask' or 'ormask'"
+        )
+    exp.mask.array[:, :] = mask
+
+    w = np.where(obs.weight > 0)
+    exp.variance.array[w] = 1.0/obs.weight[w]
+
+    if copy_psf:
+        psf_image = obs.psf.image
+        stack_psf = KernelPsf(
+            FixedKernel(
+                afw_image.ImageD(psf_image.astype(float))
+            )
+        )
+        exp.setPsf(stack_psf)
+
+    return exp
+
+
+def exp2obs(exp, copy_mask_to='ormask', store_exp=False):
+    """
+    convert an exposure to an observation.
+
+    The origin for the jacobian is set to the image center. The
+    psf image is reconstructed at the center of the image.
+
+    Optionall, the original exposure can be stored in .meta so we can reverse
+    the process and go back to an exposure, copying in any changes to the image
+    data
+
+    Parameters
+    ----------
+    exp: lsst.afw.image.ExposureF
+        The exposure
+    copy_mask_to: str
+        If set to 'ormask', the exposure mask is copied to the
+        ormask and bmask is set to zero.
+    store_exp: bool, optional
+        If set to True, store the original exposure in .meta['exposure']
+
+    Returns
+    -------
+    ngmix.Observation
+    """
+    import lsst.geom as geom
+
+    wcs = exp.getWcs()
+
+    # this is to be consistent with the coadd code
+    cen_integer, _ = get_integer_center(
+        wcs=wcs,
+        bbox=exp.getBBox(),
+    )
+    cen = geom.Point2D(cen_integer)
+
+    dm_jac = wcs.linearizePixelToSky(cen, geom.arcseconds)
+    matrix = dm_jac.getLinear().getMatrix()
+
+    dims = exp.image.array.shape
+    jrow, jcol = (np.array(dims)-1)/2
```

Comment:
there is no rounding there

Reply:
You get the integer center of the image at line 634. with `get_integer_center`. You then map this back to the sky to get the jacobian matrix. Then you set the center of the ngmix jacobian to `(np.array(dims)-1)/2`. 

This means that for an even-sized image (say 200 pixels), you would get out some integer for the center (galsim would give 99 in zero-indexed coordinates since it rounds down) but then set the ngmix jacobian center to 99.5. 

I think we the jacobian centers set at the same location?

Reply:
I was thinking this won't matter since we always are are going to convert back to an exp and it will need its bbox/wcs/etc already set.

But if we want to make it consistent I think we can do `cen.x - bbox.minx` or whatever

Reply:
`cen.x - bbox.beginX`

Reply:
Yes let's make it consistent. Agreed it likely won't matter.

Reply:
done